### PR TITLE
feat(web/ Vue): added focusshortcuts, autofocus, addonbefore, addonafter

### PR DESCRIPTION
**Pull request by mohdashraf010897**
**Description** :
- Support focusShortcuts prop to focus input field using custom keyboard shortcuts.
- support autoFocus prop for the input search field.
- Support addonBefore & addonAfter (props/ slots) to render custom jsx/ HTML markup before & after the input search field.
- Support expandSuggestionsContainer prop to expand/ limit the suggestions dropdown container width to the whole input wrapper/ input field.

**Libraries Affected:**` (hotkeys-js as a direct dependency)`

-  ReactiveSearch - web : DataSearch & CategorySearch

[Notion Link](https://www.notion.so/appbase/Support-searchbox-focus-with-keyboard-shortcut-3010b391595d4d0495a65e99f9591f9a)

[Docs PR Link](https://github.com/appbaseio/Docs/pull/193)

[**_`NEW`_** Loom Demo: ReactiveSearch - **_web_**-  focusShortcuts with/ without combinations](https://www.loom.com/share/7daf6fb4392e497bb13243eb337ce189)

[**_`NEW`_** Loom Demo: ReactiveSearch - **_vue_**-  focusShortcuts with/ without combinations](https://www.loom.com/share/ee814755c4cd4e98b8f92dbca54138d5)

[**_`OLD`_** Loom Demo: ReactiveSearch - **_web_**-  focusShortcuts, autoFocus, addonBefore, addonAfter, expandSuggestionsContainer](https://www.loom.com/share/1d77603df8a440708d7900af1312a82d)

[**_`OLD`_** Loom Demo: ReactiveSearch - **_Vue_**-  focusShortcuts, autoFocus, addonBefore, addonAfter, expandSuggestionsContainer](https://www.loom.com/share/0861e477deff4b94a7928b82d384a837)

### CODE REVIEW

**Review by bietkul — APPROVED** [review]
lgtm 👍

File: packages/vue/src/components/search/DataSearch.jsx [hunk 1]
```diff
@@ -262,6 +278,27 @@ const DataSearch = {
...
 		}
 	},
+	mounted() {
+		document.addEventListener('keydown', this.onKeyDown);
+		const { focusShortcuts } = this.$props;
+		// dynamically import hotkey-js
+		if (!isEmpty(focusShortcuts)) {
+			this.hotKeyCombinationsUsed = isHotkeyCombinationUsed(focusShortcuts);
+
+			if (this.hotKeyCombinationsUsed) {
```
Comment (bietkul) [inline packages/vue/src/components/search/DataSearch.jsx]: @mohdashraf010897 You don't need to import it dynamically. Import at the top. Fix it in `react` too. If `hot-keys` can handle the single keys then just use that and remove your custom logic to keep the bundle size minimum.

**Review by bietkul — CHANGES_REQUESTED** [review]

**Comment by bietkul** [comment]
@mohdashraf010897 Can we support special keys like `cmd`, `ctrl`, `shift` etc?

**Comment by mohdashraf010897** [comment]
> @mohdashraf010897 Can we support special keys like `cmd`, `ctrl`, `shift` etc?

We now support modifier keys as well
https://www.loom.com/share/86319a658acd481db95fb662678375d5

File: packages/vue/src/utils/index.js [hunk 2]
```diff
@@ -125,6 +125,81 @@ export const getCamelCase = (str = '') => {
...
+	return /^-?\d+$/.test(value);
+}
+
+// check if passed shortcut a key combination
+export function isHotkeyCombination(hotkey) {
+	return typeof hotkey === 'string' && hotkey.indexOf('+') !== -1;
+}
+
+// parse focusshortcuts array for key combinations
+export function isHotkeyCombinationUsed(focusShortcuts) {
```
Comment (bietkul) [inline packages/vue/src/utils/index.js]: @mohdashraf010897 Remove unused methods

File: packages/vue/src/utils/index.js [hunk 3]
```diff
@@ -125,6 +125,81 @@ export const getCamelCase = (str = '') => {
...
 };
 
+export const isEmpty = val => !(val && val.length && Object.keys(val).length);
+
+export function isNumeric(value) {
+	return /^-?\d+$/.test(value);
+}
+
+// check if passed shortcut a key combination
+export function isHotkeyCombination(hotkey) {
```
Comment (bietkul) [inline packages/vue/src/utils/index.js:135]: @mohdashraf010897 Remove if not used

File: packages/web/src/utils/index.js [hunk 4]
```diff
@@ -169,3 +169,80 @@ export const getPopularSuggestionsComponent = (data = {}, props = {}) => {
...
+	return /^-?\d+$/.test(value);
+}
+
+// check if passed shortcut a key combination
+export function isHotkeyCombination(hotkey) {
+	return typeof hotkey === 'string' && hotkey.indexOf('+') !== -1;
+}
+
+// parse focusshortcuts array for key combinations
+export function isHotkeyCombinationUsed(focusShortcuts) {
```
Comment (bietkul) [inline packages/web/src/utils/index.js]: @mohdashraf010897 remove if not used

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/utils/index.js (hunk 4 above)
Comment (mohdashraf010897) [reply packages/web/src/utils/index.js]: removed

File: packages/vue/src/utils/index.js (hunk 3 above)
Comment (mohdashraf010897) [reply packages/vue/src/utils/index.js:135]: used

File: packages/vue/src/utils/index.js (hunk 2 above)
Comment (mohdashraf010897) [reply packages/vue/src/utils/index.js]: removed

**Review by bietkul — APPROVED** [review]

**Review by bietkul — CHANGES_REQUESTED** [review]
@mohdashraf010897 FocusShortcuts doesn't work in Vue at the first attempt.
There are issues with `autoFocus` when we switch stories.

**Pull request merged**